internal/cephfs/clone.go

```
@@ -0,0 +1,111 @@
+	cephFSCloneCompleted = "complete"
+)
+
+func createCloneFromSubvolume(ctx context.Context, volID, cloneID volumeID, volOpt, parentvolOpt *volumeOptions, cr *util.Credentials) error {
+	snapshotID := cloneID
+	err := createSnapshot(ctx, parentvolOpt, cr, snapshotID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to create snapshot %s %v"), snapshotID, err)
+		return err
+	}
+	var (
+		// if protectErr is not nil we will delete the snapshot as the protect fails
+		protectErr error
+		// if cloneErr is not nil we will unprotect the snapshot and delete the snapshot
+		cloneErr error
+	)
+	defer func() {
+		if protectErr != nil {
+			err = deleteSnapshot(ctx, parentvolOpt, cr, snapshotID, volID)
+			if err != nil {
+				klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapshotID, err)
+			}
+		}
+
+		if cloneErr != nil {
+			if err = unprotectSnapshot(ctx, parentvolOpt, cr, snapshotID, volID); err != nil {
+				klog.Errorf(util.Log(ctx, "failed to unprotect snapshot %s %v"), snapshotID, err)
+			}
+			if err = deleteSnapshot(ctx, parentvolOpt, cr, snapshotID, volID); err != nil {
+				klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapshotID, err)
+			}
+		}
+	}()
+	protectErr = protectSnapshot(ctx, parentvolOpt, cr, snapshotID, volID)
+	if protectErr != nil {
+		klog.Errorf(util.Log(ctx, "failed to protect snapshot %s %v"), snapshotID, protectErr)
+		return protectErr
+	}
+
+	protectErr = cloneSnapshot(ctx, parentvolOpt, cr, volID, snapshotID, cloneID, volOpt)
+	if protectErr != nil {
+		klog.Errorf(util.Log(ctx, "failed to clone snapshot %s %s to %s %v"), volID, snapshotID, cloneID, cloneErr)
+		return protectErr
+	}
+
+	// returning cloneInProgress as we cannot do any undo operations if
+	// getcloneInfo fails, Even we are not checking the clone status here, just
+	// returning the error.
+	return ErrCloneInProgress{err: fmt.Errorf("clone is in progress for %v", cloneID)}
```

We should return an internal error if clone info fails? Clone from volume will always take 2 Create calls otherwise